## First View & Data Preparation

Let's look at what our data consist of and make a little preparation like encoding and outliers detection

In [149]:
import pandas as pd
import numpy as np
import sklearn as sk
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

In [150]:
data = pd.read_csv("data.csv")
prepared_data = data.copy()

In [151]:
data

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0     7590-VHVEG  Female              0     Yes         No       1   
1     5575-GNVDE    Male              0      No         No      34   
2     3668-QPYBK    Male              0      No         No       2   
3     7795-CFOCW    Male              0      No         No      45   
4     9237-HQITU  Female              0      No         No       2   
...          ...     ...            ...     ...        ...     ...   
7038  6840-RESVB    Male              0     Yes        Yes      24   
7039  2234-XADUH  Female              0     Yes        Yes      72   
7040  4801-JZAZL  Female              0     Yes        Yes      11   
7041  8361-LTMKD    Male              1     Yes         No       4   
7042  3186-AJIEK    Male              0      No         No      66   

     PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
0              No  No phone service             DSL             No  ...   
1             Yes                No             DSL            Yes  ...   
2             Yes                No             DSL            Yes  ...   
3              No  No phone service             DSL            Yes  ...   
4             Yes                No     Fiber optic             No  ...   
...           ...               ...             ...            ...  ...   
7038          Yes               Yes             DSL            Yes  ...   
7039          Yes               Yes     Fiber optic             No  ...   
7040           No  No phone service             DSL            Yes  ...   
7041          Yes               Yes     Fiber optic             No  ...   
7042          Yes                No     Fiber optic            Yes  ...   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No          No          No              No  Month-to-month   
1                 Yes          No          No              No        One year   
2                  No          No          No              No  Month-to-month   
3                 Yes         Yes          No              No        One year   
4                  No          No          No              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7038              Yes         Yes         Yes             Yes        One year   
7039              Yes          No         Yes             Yes        One year   
7040               No          No          No              No  Month-to-month   
7041               No          No          No              No  Month-to-month   
7042              Yes         Yes         Yes             Yes        Two year   

     PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0                 Yes           Electronic check          29.85         29.85   
1                  No               Mailed check          56.95        1889.5   
2                 Yes               Mailed check          53.85        108.15   
3                  No  Bank transfer (automatic)          42.30       1840.75   
4                 Yes           Electronic check          70.70        151.65   
...               ...                        ...            ...           ...   
7038              Yes               Mailed check          84.80        1990.5   
7039              Yes    Credit card (automatic)         103.20        7362.9   
7040              Yes           Electronic check          29.60        346.45   
7041              Yes               Mailed check          74.40         306.6   
7042              Yes  Bank transfer (automatic)         105.65        6844.5   

     Churn  
0       No  
1       No  
2      Yes  
3       No  
4      Yes  
...    ...  
7038    No  
7039    No  
7040    No  
7041   Yes  
7042    No  

[7043 rows x 21 columns]

In [152]:
print(data.shape)

(7043, 21)


In [153]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


As we see we have a dataset of 7043 objects and there are many categorical features in the data.

Let's check how many empty cells has our data

In [154]:
data.isna().sum().sum()

0

# Encoding 

At least we don't have to think about missing data. Now let's go through every feature and think what does it actually mean. Then encode it for the future analysis.

### customerID

Not an interesting feature for us. Just a number for a user of a system

In [155]:
data["customerID"].head()

0    7590-VHVEG
1    5575-GNVDE
2    3668-QPYBK
3    7795-CFOCW
4    9237-HQITU
Name: customerID, dtype: object

### 1. gender 

Pretty obvious. We will encode it just like a binary variable. But before we just need to ensure that in this column there are only 2 classes (makes much sense in the current world...)

In [156]:
le = LabelEncoder()

prepared_data["gender"] = le.fit_transform(data["gender"]).astype(float)

According to our labels 0 means Female when 1 is Male

### 2. SeniorCitizen

Just an indicator whether the customer is a senior citizen or not. It already has integers values so all is good.

In [157]:
prepared_data["SeniorCitizen"] = prepared_data["SeniorCitizen"].astype(float)
prepared_data["SeniorCitizen"].value_counts()

SeniorCitizen
0.0    5901
1.0    1142
Name: count, dtype: int64

### Partner & Dependents (3 and 4)

Two features which show if a customer has a Partner and Dependents. Both has tho categories (Yes, No)

In [158]:
data["Partner"].value_counts()

Partner
No     3641
Yes    3402
Name: count, dtype: int64

In [159]:
data["Dependents"].value_counts()

Dependents
No     4933
Yes    2110
Name: count, dtype: int64

In [160]:
le = LabelEncoder()

le.fit(data["Partner"])
le.classes_

array(['No', 'Yes'], dtype=object)

In [161]:
prepared_data["Partner"] = le.transform(data["Partner"]).astype(float)
prepared_data["Dependents"] = le.transform(data["Dependents"]).astype(float)

### 5. tenure

It's a number of months the customer has stayed with the company. Already an integer number

In [162]:
prepared_data["tenure"] = data["tenure"].astype(float)
prepared_data["tenure"].head()

0     1.0
1    34.0
2     2.0
3    45.0
4     2.0
Name: tenure, dtype: float64

### 6. PhoneService

Whether the customer has a phone service or not (Yes, No)

In [163]:
le = LabelEncoder()

le.fit(data["PhoneService"])

prepared_data["PhoneService"] = le.transform(data["PhoneService"]).astype(float)

### 7. MultipleLines 

Whether the customer has multiple lines or not (Yes, No, No phone service). It looks like a common situation for OneHotEncoder unless we notice that if we encode it with One Hot, we will have two features that will mean actually the same thing. We will have feature "Phone Service" and a category column "MultipleLines_No_Phone_service". They will be "binary inverse" (where the first one has zeros, the second one has ones).

So we can just encode this feature as follows. Where MultipleLines is "Yes" we set 1 and 0 otherwise

In [164]:
prepared_data.loc[data["MultipleLines"] == "Yes", "MultipleLines"] = 1
prepared_data.loc[data["MultipleLines"] != "Yes", "MultipleLines"] = 0
prepared_data["MultipleLines"] = prepared_data["MultipleLines"].astype(float)

### 8. InternetService

Customer’s internet service provider (DSL, Fiber optic, No). Here we will use One Code Encoder.

In [165]:
enc = OneHotEncoder(sparse_output=False)

one_hot_encoded = enc.fit_transform(data[["InternetService"]])

one_hot_df = pd.DataFrame(one_hot_encoded,
                          columns=enc.get_feature_names_out(["InternetService"]))

prepared_data = pd.concat([prepared_data, one_hot_df], axis=1)

prepared_data = prepared_data.drop(["InternetService"], axis=1)

### 9. OnlineSecurity

Whether the customer has online security or not (Yes, No, No internet service).

Like in the situation with Multiple Lines if we use One Hot we will receive 2 columns which have the same information. So we will just set 1 where OnlineSecurity is 1 and 0 otherwise

In [166]:
prepared_data.loc[data["OnlineSecurity"] == "Yes", "OnlineSecurity"] = 1
prepared_data.loc[data["OnlineSecurity"] != "Yes", "OnlineSecurity"] = 0
prepared_data["OnlineSecurity"] = prepared_data["OnlineSecurity"].astype(float)

#### 10. OnlineBackup 

Whether the customer has online backup or not (Yes, No, No internet service).

#### 11. DeviceProtection

Whether the customer has device protection or not (Yes, No, No internet service)

#### 12. TechSupport

Whether the customer has tech support or not (Yes, No, No internet service)

#### 13. StreamingTV

Whether the customer has streaming TV or not (Yes, No, No internet service)

#### 14. StreamingMovies

Whether the customer has streaming movies or not (Yes, No, No internet service)

Every of these features have the same situation as OnlineSecurity and MultipleLines.

In [167]:
prepared_data.loc[data["OnlineBackup"] == "Yes", "OnlineBackup"] = 1
prepared_data.loc[data["OnlineBackup"] != "Yes", "OnlineBackup"] = 0
prepared_data["OnlineBackup"] = prepared_data["OnlineBackup"].astype(float)

prepared_data.loc[data["DeviceProtection"] == "Yes", "DeviceProtection"] = 1
prepared_data.loc[data["DeviceProtection"] != "Yes", "DeviceProtection"] = 0
prepared_data["DeviceProtection"] = prepared_data["DeviceProtection"].astype(float)

prepared_data.loc[data["TechSupport"] == "Yes", "TechSupport"] = 1
prepared_data.loc[data["TechSupport"] != "Yes", "TechSupport"] = 0
prepared_data["TechSupport"] = prepared_data["TechSupport"].astype(float)

prepared_data.loc[data["StreamingTV"] == "Yes", "StreamingTV"] = 1
prepared_data.loc[data["StreamingTV"] != "Yes", "StreamingTV"] = 0
prepared_data["StreamingTV"] = prepared_data["StreamingTV"].astype(float)

prepared_data.loc[data["StreamingMovies"] == "Yes", "StreamingMovies"] = 1
prepared_data.loc[data["StreamingMovies"] != "Yes", "StreamingMovies"] = 0
prepared_data["StreamingMovies"] = prepared_data["StreamingMovies"].astype(float)

### 15. Contract

The contract term of the customer (Month-to-month, One year, Two year).

Just a common situation for One Hot

In [168]:
enc = OneHotEncoder(sparse_output=False)

one_hot_encoded = enc.fit_transform(data[["Contract"]])

one_hot_df = pd.DataFrame(one_hot_encoded,
                          columns=enc.get_feature_names_out(["Contract"]))

prepared_data = pd.concat([prepared_data, one_hot_df], axis=1)

prepared_data = prepared_data.drop(["Contract"], axis=1)

In [169]:
prepared_data.head()

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG     0.0            0.0      1.0         0.0     1.0   
1  5575-GNVDE     1.0            0.0      0.0         0.0    34.0   
2  3668-QPYBK     1.0            0.0      0.0         0.0     2.0   
3  7795-CFOCW     1.0            0.0      0.0         0.0    45.0   
4  9237-HQITU     0.0            0.0      0.0         0.0     2.0   

   PhoneService  MultipleLines  OnlineSecurity  OnlineBackup  ...  \
0           0.0            0.0             0.0           1.0  ...   
1           1.0            0.0             1.0           0.0  ...   
2           1.0            0.0             1.0           1.0  ...   
3           0.0            0.0             1.0           0.0  ...   
4           1.0            0.0             0.0           0.0  ...   

               PaymentMethod  MonthlyCharges  TotalCharges  Churn  \
0           Electronic check           29.85         29.85     No   
1               Mailed check           56.95        1889.5     No   
2               Mailed check           53.85        108.15    Yes   
3  Bank transfer (automatic)           42.30       1840.75     No   
4           Electronic check           70.70        151.65    Yes   

  InternetService_DSL InternetService_Fiber optic  InternetService_No  \
0                 1.0                         0.0                 0.0   
1                 1.0                         0.0                 0.0   
2                 1.0                         0.0                 0.0   
3                 1.0                         0.0                 0.0   
4                 0.0                         1.0                 0.0   

  Contract_Month-to-month Contract_One year  Contract_Two year  
0                     1.0               0.0                0.0  
1                     0.0               1.0                0.0  
2                     1.0               0.0                0.0  
3                     0.0               1.0                0.0  
4                     1.0               0.0                0.0  

[5 rows x 25 columns]

### 16. PaperlessBilling

Whether the customer has paperless billing or not (Yes, No)

In [170]:
le = LabelEncoder()

le.fit(data["PaperlessBilling"])
print(le.classes_)

prepared_data["PaperlessBilling"] = le.transform(data["PaperlessBilling"]).astype(float)

['No' 'Yes']


### 17. PaymentMethod

The customer’s payment method (Electronic check, Mailed check, Bank transfer (automatic), Credit card

In [171]:
enc = OneHotEncoder(sparse_output=False)

one_hot_encoded = enc.fit_transform(data[["PaymentMethod"]])

one_hot_df = pd.DataFrame(one_hot_encoded,
                          columns=enc.get_feature_names_out(["PaymentMethod"]))

prepared_data = pd.concat([prepared_data, one_hot_df], axis=1)

prepared_data = prepared_data.drop(["PaymentMethod"], axis=1)

In [172]:
prepared_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 28 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   customerID                               7043 non-null   object 
 1   gender                                   7043 non-null   float64
 2   SeniorCitizen                            7043 non-null   float64
 3   Partner                                  7043 non-null   float64
 4   Dependents                               7043 non-null   float64
 5   tenure                                   7043 non-null   float64
 6   PhoneService                             7043 non-null   float64
 7   MultipleLines                            7043 non-null   float64
 8   OnlineSecurity                           7043 non-null   float64
 9   OnlineBackup                             7043 non-null   float64
 10  DeviceProtection                         7043 no

In [173]:
prepared_data

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0     7590-VHVEG     0.0            0.0      1.0         0.0     1.0   
1     5575-GNVDE     1.0            0.0      0.0         0.0    34.0   
2     3668-QPYBK     1.0            0.0      0.0         0.0     2.0   
3     7795-CFOCW     1.0            0.0      0.0         0.0    45.0   
4     9237-HQITU     0.0            0.0      0.0         0.0     2.0   
...          ...     ...            ...      ...         ...     ...   
7038  6840-RESVB     1.0            0.0      1.0         1.0    24.0   
7039  2234-XADUH     0.0            0.0      1.0         1.0    72.0   
7040  4801-JZAZL     0.0            0.0      1.0         1.0    11.0   
7041  8361-LTMKD     1.0            1.0      1.0         0.0     4.0   
7042  3186-AJIEK     1.0            0.0      0.0         0.0    66.0   

      PhoneService  MultipleLines  OnlineSecurity  OnlineBackup  ...  \
0              0.0            0.0             0.0           1.0  ...   
1              1.0            0.0             1.0           0.0  ...   
2              1.0            0.0             1.0           1.0  ...   
3              0.0            0.0             1.0           0.0  ...   
4              1.0            0.0             0.0           0.0  ...   
...            ...            ...             ...           ...  ...   
7038           1.0            1.0             1.0           0.0  ...   
7039           1.0            1.0             0.0           1.0  ...   
7040           0.0            0.0             1.0           0.0  ...   
7041           1.0            1.0             0.0           0.0  ...   
7042           1.0            0.0             1.0           0.0  ...   

      InternetService_DSL  InternetService_Fiber optic  InternetService_No  \
0                     1.0                          0.0                 0.0   
1                     1.0                          0.0                 0.0   
2                     1.0                          0.0                 0.0   
3                     1.0                          0.0                 0.0   
4                     0.0                          1.0                 0.0   
...                   ...                          ...                 ...   
7038                  1.0                          0.0                 0.0   
7039                  0.0                          1.0                 0.0   
7040                  1.0                          0.0                 0.0   
7041                  0.0                          1.0                 0.0   
7042                  0.0                          1.0                 0.0   

      Contract_Month-to-month  Contract_One year  Contract_Two year  \
0                         1.0                0.0                0.0   
1                         0.0                1.0                0.0   
2                         1.0                0.0                0.0   
3                         0.0                1.0                0.0   
4                         1.0                0.0                0.0   
...                       ...                ...                ...   
7038                      0.0                1.0                0.0   
7039                      0.0                1.0                0.0   
7040                      1.0                0.0                0.0   
7041                      1.0                0.0                0.0   
7042                      0.0                0.0                1.0   

     PaymentMethod_Bank transfer (automatic)  \
0                                        0.0   
1                                        0.0   
2                                        0.0   
3                                        1.0   
4                                        0.0   
...                                      ...   
7038                                     0.0   
7039                                     0.0   
7040                                     0.0   
7041  

### 18. MonthlyCharges

The amount charged to the customer monthly. Already in float

In [174]:
prepared_data["MonthlyCharges"].head()

0    29.85
1    56.95
2    53.85
3    42.30
4    70.70
Name: MonthlyCharges, dtype: float64

### 19. TotalCharges

The total amount charged to the customer. For some reason this feature is considered by pandas like "object". After quick research it can be found that there are 11 rows in data where TotalCharges feature is filled like empty string ' '

In [175]:
data[data["TotalCharges"] == ' ']

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
488   4472-LVYGI  Female              0     Yes        Yes       0   
753   3115-CZMZD    Male              0      No        Yes       0   
936   5709-LVOEQ  Female              0     Yes        Yes       0   
1082  4367-NUYAO    Male              0     Yes        Yes       0   
1340  1371-DWPAZ  Female              0     Yes        Yes       0   
3331  7644-OMVMY    Male              0     Yes        Yes       0   
3826  3213-VVOLG    Male              0     Yes        Yes       0   
4380  2520-SGTTA  Female              0     Yes        Yes       0   
5218  2923-ARZLG    Male              0     Yes        Yes       0   
6670  4075-WKNIU  Female              0     Yes        Yes       0   
6754  2775-SEFEE    Male              0      No        Yes       0   

     PhoneService     MultipleLines InternetService       OnlineSecurity  ...  \
488            No  No phone service             DSL                  Yes  ...   
753           Yes                No              No  No internet service  ...   
936           Yes                No             DSL                  Yes  ...   
1082          Yes               Yes              No  No internet service  ...   
1340           No  No phone service             DSL                  Yes  ...   
3331          Yes                No              No  No internet service  ...   
3826          Yes               Yes              No  No internet service  ...   
4380          Yes                No              No  No internet service  ...   
5218          Yes                No              No  No internet service  ...   
6670          Yes               Yes             DSL                   No  ...   
6754          Yes               Yes             DSL                  Yes  ...   

         DeviceProtection          TechSupport          StreamingTV  \
488                   Yes                  Yes                  Yes   
753   No internet service  No internet service  No internet service   
936                   Yes                   No                  Yes   
1082  No internet service  No internet service  No internet service   
1340                  Yes                  Yes                  Yes   
3331  No internet service  No internet service  No internet service   
3826  No internet service  No internet service  No internet service   
4380  No internet service  No internet service  No internet service   
5218  No internet service  No internet service  No internet service   
6670                  Yes                  Yes                  Yes   
6754                   No                  Yes                   No   

          StreamingMovies  Contract PaperlessBilling  \
488                    No  Two year              Yes   
753   No internet service  Two year               No   
936                   Yes  Two year               No   
1082  No internet service  Two year               No   
1340                   No  Two year               No   
3331  No internet service  Two year               No   
3826  No internet service  Two year               No   
4380  No internet service  Two year               No   
5218  No internet service  One year              Yes   
6670                   No  Two year               No   
6754                   No  Two year              Yes   

                  PaymentMethod MonthlyCharges  TotalCharges Churn  
488   Bank transfer (automatic)          52.55                  No  
753                Mailed check          20.25                  No  
936                Mailed check          80.85                  No  
1082               Mailed check          25.75                  No  
1340    Credit card (automatic)          56.05                  No  
3331               Mailed check          19.85                  No  
3826               Mailed check          25.35                  No  
4380               Mailed check          20.00                  No  
5218               Mailed check          19.7

If we think why is that we can understand that these are customers which have just joined to the company (tenure is equal to zero) and they have not paid yet.

To ensure we can check how many customers have tenure equal to zero

In [176]:
data[data["tenure"] == 0]

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
488   4472-LVYGI  Female              0     Yes        Yes       0   
753   3115-CZMZD    Male              0      No        Yes       0   
936   5709-LVOEQ  Female              0     Yes        Yes       0   
1082  4367-NUYAO    Male              0     Yes        Yes       0   
1340  1371-DWPAZ  Female              0     Yes        Yes       0   
3331  7644-OMVMY    Male              0     Yes        Yes       0   
3826  3213-VVOLG    Male              0     Yes        Yes       0   
4380  2520-SGTTA  Female              0     Yes        Yes       0   
5218  2923-ARZLG    Male              0     Yes        Yes       0   
6670  4075-WKNIU  Female              0     Yes        Yes       0   
6754  2775-SEFEE    Male              0      No        Yes       0   

     PhoneService     MultipleLines InternetService       OnlineSecurity  ...  \
488            No  No phone service             DSL                  Yes  ...   
753           Yes                No              No  No internet service  ...   
936           Yes                No             DSL                  Yes  ...   
1082          Yes               Yes              No  No internet service  ...   
1340           No  No phone service             DSL                  Yes  ...   
3331          Yes                No              No  No internet service  ...   
3826          Yes               Yes              No  No internet service  ...   
4380          Yes                No              No  No internet service  ...   
5218          Yes                No              No  No internet service  ...   
6670          Yes               Yes             DSL                   No  ...   
6754          Yes               Yes             DSL                  Yes  ...   

         DeviceProtection          TechSupport          StreamingTV  \
488                   Yes                  Yes                  Yes   
753   No internet service  No internet service  No internet service   
936                   Yes                   No                  Yes   
1082  No internet service  No internet service  No internet service   
1340                  Yes                  Yes                  Yes   
3331  No internet service  No internet service  No internet service   
3826  No internet service  No internet service  No internet service   
4380  No internet service  No internet service  No internet service   
5218  No internet service  No internet service  No internet service   
6670                  Yes                  Yes                  Yes   
6754                   No                  Yes                   No   

          StreamingMovies  Contract PaperlessBilling  \
488                    No  Two year              Yes   
753   No internet service  Two year               No   
936                   Yes  Two year               No   
1082  No internet service  Two year               No   
1340                   No  Two year               No   
3331  No internet service  Two year               No   
3826  No internet service  Two year               No   
4380  No internet service  Two year               No   
5218  No internet service  One year              Yes   
6670                   No  Two year               No   
6754                   No  Two year              Yes   

                  PaymentMethod MonthlyCharges  TotalCharges Churn  
488   Bank transfer (automatic)          52.55                  No  
753                Mailed check          20.25                  No  
936                Mailed check          80.85                  No  
1082               Mailed check          25.75                  No  
1340    Credit card (automatic)          56.05                  No  
3331               Mailed check          19.85                  No  
3826               Mailed check          25.35                  No  
4380               Mailed check          20.00                  No  
5218               Mailed check          19.7

We received the same subset of customers

It is obvious that in this case the TotalCharges feature is just zero. So

In [177]:
prepared_data.loc[data["TotalCharges"] == ' ', "TotalCharges"] = 0
prepared_data["TotalCharges"] = prepared_data["TotalCharges"].astype(float)

### 20. Churn

Whether the customer churned or not (Yes or No)

We label it as 1 and 0 and push to the right side of our DataFrame

In [178]:
le = LabelEncoder()

le.fit(data["Churn"])
print(le.classes_)
new_Churn = le.transform(data["Churn"]).astype(float)

prepared_data = prepared_data.drop(["Churn"], axis=1)

prepared_data["Churn"] = new_Churn

['No' 'Yes']


Now all seems to be good and we have encoded data to analyse

In [179]:
prepared_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 28 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   customerID                               7043 non-null   object 
 1   gender                                   7043 non-null   float64
 2   SeniorCitizen                            7043 non-null   float64
 3   Partner                                  7043 non-null   float64
 4   Dependents                               7043 non-null   float64
 5   tenure                                   7043 non-null   float64
 6   PhoneService                             7043 non-null   float64
 7   MultipleLines                            7043 non-null   float64
 8   OnlineSecurity                           7043 non-null   float64
 9   OnlineBackup                             7043 non-null   float64
 10  DeviceProtection                         7043 no

In [180]:
prepared_data.head(5)

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG     0.0            0.0      1.0         0.0     1.0   
1  5575-GNVDE     1.0            0.0      0.0         0.0    34.0   
2  3668-QPYBK     1.0            0.0      0.0         0.0     2.0   
3  7795-CFOCW     1.0            0.0      0.0         0.0    45.0   
4  9237-HQITU     0.0            0.0      0.0         0.0     2.0   

   PhoneService  MultipleLines  OnlineSecurity  OnlineBackup  ...  \
0           0.0            0.0             0.0           1.0  ...   
1           1.0            0.0             1.0           0.0  ...   
2           1.0            0.0             1.0           1.0  ...   
3           0.0            0.0             1.0           0.0  ...   
4           1.0            0.0             0.0           0.0  ...   

   InternetService_Fiber optic  InternetService_No  Contract_Month-to-month  \
0                          0.0                 0.0                      1.0   
1                          0.0                 0.0                      0.0   
2                          0.0                 0.0                      1.0   
3                          0.0                 0.0                      0.0   
4                          1.0                 0.0                      1.0   

   Contract_One year  Contract_Two year  \
0                0.0                0.0   
1                1.0                0.0   
2                0.0                0.0   
3                1.0                0.0   
4                0.0                0.0   

   PaymentMethod_Bank transfer (automatic)  \
0                                      0.0   
1                                      0.0   
2                                      0.0   
3                                      1.0   
4                                      0.0   

   PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
0                                    0.0                             1.0   
1                                    0.0                             0.0   
2                                    0.0                             0.0   
3                                    0.0                             0.0   
4                                    0.0                             1.0   

   PaymentMethod_Mailed check  Churn  
0                         0.0    0.0  
1                         1.0    0.0  
2                         1.0    1.0  
3                         0.0    0.0  
4                         0.0    1.0  

[5 rows x 28 columns]

In [181]:
prepared_data.to_csv("prepared_data.csv", index=False)